SingularityService/src/test/java/com/hubspot/singularity/scheduler/SingularityUsageTest.java

```diff
@@ -688,6 +706,249 @@ public void itLimitsTheNumberOfTaskCleanupsToCreate() {
+      mesosClient.setSlaveMetricsSnapshot(
+          "host1",
+          new MesosSlaveMetricsSnapshotObject(0, 0, 0, 10.0, 0, 0, 0, 0, 0, 0, 0, 0, 10.0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 100000, 0, 1000, 0, 0, 0, 10, 0, 0, 0, 0)
+      );
+
+      usagePoller.runActionOnPoll();
+
+      System.out.println(taskManager.getCleanupTaskIds().toString());
+
+      // First task is not cleaned up, due to relatively high utilization.
+      Assertions.assertFalse(taskManager.getTaskCleanup(task1.getId()).isPresent());
+
+      // Second task is cleaned up, in order to reach desired memory threshold.
+      Assertions.assertEquals(TaskCleanupType.REBALANCE_MEMORY_USAGE, taskManager.getTaskCleanup(task2.getId()).get().getCleanupType());
+
+      // Third task is not cleaned up, as doing so will not reach the desired memory threshold given available shuffles.
+      Assertions.assertFalse(taskManager.getTaskCleanup(task3.getId()).isPresent());
+
+      // Fourth task is cleaned up, as it has the lowest utilization.
+      Assertions.assertEquals(TaskCleanupType.REBALANCE_MEMORY_USAGE, taskManager.getTaskCleanup(task4.getId()).get().getCleanupType());
+    } finally {
+      configuration.setShuffleTasksForOverloadedSlaves(false);
+    }
+  }
+
+  protected void scheduleTask(String rqid, double requiredCpus, double requiredMemoryMb) {
+    Resources rs = new Resources(requiredCpus, requiredMemoryMb, 0);
+    SingularityRequest rq = new SingularityRequestBuilder(rqid, RequestType.WORKER)
+        .setInstances(Optional.of(1))
+        .build();
+
+    SingularityDeploy dp = new SingularityDeployBuilder(rq.getId(), "deployId1")
+        .setCommand(Optional.of("sleep 100"))
+        .setResources(Optional.of(rs))
+        .build();
+
+    SingularityDeployMarker marker = new SingularityDeployMarker(dp.getRequestId(), dp.getId(), System.currentTimeMillis(), Optional.empty(), Optional.empty());
+    SingularityPendingRequest pending = new SingularityPendingRequest(
+        rq.getId(),
+        dp.getId(),
+        System.currentTimeMillis(),
+        Optional.empty(),
+        SingularityPendingRequest.PendingType.UPDATED_REQUEST,
+        Optional.empty(),
+        Optional.empty()
+    );
+
+    requestManager.activate(rq, SingularityRequestHistory.RequestHistoryType.CREATED, System.currentTimeMillis(), Optional.empty(), Optional.empty());
+    deployManager.saveDeploy(rq, marker, dp);
+    deployManager.deletePendingDeploy(marker.getRequestId());
+    deployManager.saveDeployResult(marker, Optional.of(dp), new SingularityDeployResult(DeployState.SUCCEEDED));
+    deployManager.saveNewRequestDeployState(new SingularityRequestDeployState(marker.getRequestId(), Optional.of(marker), Optional.empty()));
+
+    requestManager.addToPendingQueue(pending);
+    scheduler.drainPendingQueue();
+  }
+
+  protected Map<String, Map<String, SingularityTaskId>> getTaskIdMapByHostByRequest() {
```

Comment:
I don't know that we really need to build a map of this here. It seems in each case you are just using the getActiveTaskIds to then loop through all of them and get them into an active/starting state which can then be viewed by the usage poller. Can probably just do a for loop on taskManager.getActiveTaskIds() after maybe validating the size of the list

Reply:
I feel the map is more intuitive than looping through the task ids and matching host/request names, especially when working with multiple hosts, but it's definitely not something that needs to be built. I'd prefer to leave it, but will remove it if you prefer a simpler structure.